otherlibs/stdune/src/path.ml

```diff
@@ -1220,7 +1208,68 @@ let source s = in_source_tree s
 
 let build s = in_build_dir s
 
-module Table = Hashtbl.Make (T)
+module Table = struct
+  type 'a t =
+    { source : 'a Source0.Table.t
+    ; build : 'a Build.Table.t
+    ; external_ : 'a External.Table.t
+    }
+
+  let create size =
+    { source = Source0.Table.create size
+    ; build = Build.Table.create size
+    ; external_ = External.Table.create size
+    }
+
+  let clear { source; build; external_ } =
+    Source0.Table.clear source;
+    Build.Table.clear build;
+    External.Table.clear external_
+
+  let mem { source; build; external_ } key =
+    match key with
+    | In_source_tree p -> Source0.Table.mem source p
+    | In_build_dir p -> Build.Table.mem build p
+    | External p -> External.Table.mem external_ p
+
+  let set { source; build; external_ } k v =
+    match k with
+    | In_source_tree p -> Source0.Table.set source p v
+    | In_build_dir p -> Build.Table.set build p v
+    | External p -> External.Table.set external_ p v
+
+  let remove { source; build; external_ } k =
+    match k with
+    | In_source_tree p -> Source0.Table.remove source p
+    | In_build_dir p -> Build.Table.remove build p
+    | External p -> External.Table.remove external_ p
+
+  let iter { source; build; external_ } ~f =
+    Source0.Table.iter source ~f;
+    Build.Table.iter build ~f;
+    External.Table.iter external_ ~f
+
+  let find { source; build; external_ } = function
+    | In_source_tree p -> Source0.Table.find source p
+    | In_build_dir p -> Build.Table.find build p
+    | External p -> External.Table.find external_ p
+
+  let filteri_inplace { source; build; external_ } ~f =
+    Source0.Table.filteri_inplace source ~f:(fun ~key ~data ->
+        f ~key:(In_source_tree key) ~data);
```

Comment:
This is likely worse than before.

Reply:
Looks like there is just one call site (`src/dune_engine/cached_digest.ml`). Maybe we can change it to avoid the possibility of regression.

Reply:
Indeed it's a little worse. I'll tweak the API to pass an iteration function for every type of key. That should make it the same as before.

Reply:
Done. I've fixed the regression and made it even faster by doing the debug check only once.